src/io/linux.zig

```diff
@@ -960,6 +960,84 @@ pub const IO = struct {
         self.enqueue(completion);
     }
 
+    pub const INVALID_EVENT = @bitCast(u32, @as(i32, -1));
+
+    pub const EventResponse = enum {
+        listen,
+        cancel,
+    };
+
+    pub fn open_event(
+        self: *IO,
+        completion: *Completion,
+        comptime on_event: fn (*Completion) EventResponse,
+    ) !u32 {
+        // eventfd initialized with no (zero) previous write value,
+        const event_fd = os.eventfd(0, linux.EFD.CLOEXEC) catch |err| switch (err) {
+            error.SystemResources, error.SystemFdQuotaExceeded, error.ProcessFdQuotaExceeded => return error.SystemResources,
+            error.Unexpected => return error.Unexpected,
+        };
+        assert(event_fd != INVALID_EVENT);
+        errdefer os.close(event_fd);
+
+        // Start listening for notifications on this event.
+        const event = @intCast(u32, event_fd);
+        self.listen_event(event, completion, on_event);
+        return event;
+    }
+
+    fn listen_event(
+        self: *IO,
+        event: u32,
+        completion: *Completion,
+        comptime on_event: fn (*Completion) EventResponse,
+    ) void {
+        const Listen = struct {
+            var stub_value: u64 = undefined;
+
+            fn on_read(ctx: *anyopaque, _completion: *Completion, result: ReadError!usize) void {
+                const bytes = result catch unreachable; // eventfd reads should not fail.
+                assert(bytes == @sizeOf(u64));
+
+                const _event = @intCast(u32, @ptrToInt(ctx));
+                assert(_event != INVALID_EVENT);
+
+                switch (on_event(_completion)) {
+                    .listen => _completion.io.listen_event(_event, _completion, on_event),
+                    .cancel => {},
+                }
+            }
+        };
+
+        self.read(
+            *anyopaque,
+            @intToPtr(*anyopaque, event),
```

Comment:
Why does io.read take a type argument if the only type that works is `*anyopaque`?

Reply:
It only supports a pointer type, hence the use of `anyopaque` just for representing the `eventfd` as pointer.

Reply:
Oh, I saw that it doesn't cast `context` [here](https://github.com/tigerbeetledb/tigerbeetle/blob/0b15d2209f6d51f2f56e7f7a597540b78ddacdf6//src/io/linux.zig#L652), but I didn't know that there is an implicit conversion from *T to *anyopaque.